File: Graph.cpp
#include <iostream>
#include <list>
#include <vector>

using namespace std;

template<typename V, typename E>
class Graph {
public:

    class Vertex;

    class Edge;

    class Vertex {
    public:
        V value;
        int ID;
        list<Edge *> adjNodes;
        typename vector<Vertex *>::iterator vertexRef;

        Vertex() {
        }

        Vertex(V val, int ID) {
            value = val;
            this->ID = ID;
            //adjNodes = list<Edge*>();
        }

    };

    class Edge {
    public:
        E weight;
        Vertex *from;
        Vertex *to;
        typename list<Edge *>::iterator edgeRef;

        Edge(E wei, Vertex *fr, Vertex *t) {
            weight = wei;
            from = fr;
            to = t;
        }

        bool operator<(const Edge &a) const {
            return weight < a.weight;
        }
    };

    int n = 0;
    vector<Vertex *> vertices;
    list<Edge *> edges;
    list<Edge> MST;

    Graph() {
        vertices = new vector<Vertex>();
        edges = list<Edge>();
        MST = list<Edge>();
    }

    Vertex *addVertex(V val) {
        Vertex *vertex = new Vertex(val, n);
        n++;
        vertex->vertexRef = vertices.emplace(vertices.end(), vertex);
        return vertex;
    }

    Edge *addEdge(E weight, Vertex *from, Vertex *to) {
        Edge *edge = new Edge(weight, from, to);
        from->adjNodes.push_back(edge);
        // from.adjNodes.insert(from.adjNodes.end(), edge);
        edge->edgeRef = edges.insert(edges.end(), edge);
        return edge;
    }

    int inDegree(Vertex vertex) {
        int result = 0;
        for (Edge edge : edges)
            if (edge->to->value == vertex.value && edge->to->ID == vertex.ID) result++;
        return result;
    }

    int outDegree(Vertex vertex) {
        int result = 0;
        for (Edge edge : edges)
            if (edge->from->value == vertex.value && edge->from->ID == vertex.ID) result++;
        return result;
    }

    bool areAdjacent(Vertex v, Vertex u) {
        for (Edge edge: edges) {
            if ((edge.from.value == v.value && edge.to.value == u.value) ||
                (edge.from.value == u.value && edge.to.value == v.value))
                return true;
        }
        return false;
    }

    void removeEdge(Edge edge) {
        edges.erase(edge.edgeRef);
    }

    void removeVertex(Vertex vertex) {
        vertices.erase(vertex.vertexRef);

        auto it = edges.begin();

        while (it != edges.end()) {
            if ((*it).from.value == vertex.value || (*it).to.value == vertex.value) {
                it = edges.erase(it);
            } else
                it++;
        }
    }

    pair<Vertex, Vertex> findEdge(Edge edge) { // returns vertices between this edge
        for (Edge e: edges)
            if (e->from->value == edge.from->value && e->to->value == edge.to->value && e->weight == edge.weight)
                return make_pair(e->from, e->to);
        cout << "Edge was not found!!!";
        return;
    }

    bool containsEdge(Edge edge) {
        for (Edge e: edges)
            if (e->from->value == edge.from->value && e->to->value == edge.to->value && e->weight == edge.weight)
                return true;
        return false;
    }

    vector<Edge> outgoingEdges(Vertex vertex) {
        vector<Edge> result;
        for (Edge edge : edges)
            if (edge->from == vertex)
                result.push_back(edge);
        return result;
    }

    vector<Edge> incomingEdges(Vertex vertex) {
        vector<Edge> result;
        for (Edge edge : edges)
            if (edge->to == vertex)
                result.push_back(edge);
        return result;
    }

    pair<Vertex, Vertex> getEdge(Vertex u, Vertex v) { // returns vertices between this edge
        for (Edge edge : edges)
            if ((edge->from->value == u.value && edge->from->ID == u.ID)
                && (edge->to->value == v.value && edge->to->ID == v.ID))
                return make_pair(u, v);
        cout << "Edge was not found!!!";
        return;
    }

    bool empty() {
        return vertices.empty();
    }

    int numberOfVertices() {
        return vertices.size();
    }

    int numberOfEdges() {
        return edges.size();
    }

    vector<Vertex> allVertices() {
        vector<Vertex> result;
        for (Vertex v : vertices)
            result.push_back(v);
        return result;
    }

    vector<Edge> allEdges() {
        vector<Edge> result;
        for (Edge e : edges)
            result.push_back(e);
        return result;
    }


};

int main() {

    return 0;
}

File: Graph.java
import java.util.ArrayList;
import java.util.List;

public class Graph {
    public static void main(String[] args) {
        
    }
}


class Vertex<V> {
    V value;
    int ID;

    public Vertex(V value, int ID) {
        this.value = value;
        this.ID = ID;
    }
}

class WeightedGraph<V, E> {
    private class Vertex {
        V value;
        int ID;

        public Vertex(V value, int ID) {
            this.value = value;
            this.ID = ID;
        }
    }

    private class Edge {
        Vertex from, to;
        E weight;

        public Edge(Vertex from, Vertex to, E weight) {
            this.from = from;
            this.to = to;
            this.weight = weight;
        }
    }

    int n = 0;
    int numV, numE;
    private ArrayList<Vertex> vertices;
    private ArrayList<Edge> edges;

    public WeightedGraph(int numV, int numE) {
        this.numV = numV;
        this.numE = numE;
        this.vertices = new ArrayList<>(numV);
        this.edges = new ArrayList<>(numE);
    }

    public WeightedGraph() {
        this.vertices = new ArrayList<>();
        this.edges = new ArrayList<>();
    }

    public void addVertex(V value, int ID) {
        Vertex v = new Vertex(value, n++);
        vertices.add(v);
    }

    public void addEdge(Vertex from, Vertex to, E weight) {
        Edge e = new Edge(from, to, weight);
        edges.add(e);
    }

    public boolean areAdjacent(Vertex v1, Vertex v2) {
        for (Edge e : edges) {
            if ((e.from.value == v1.value && e.to.value == v2.value) ||
                    (e.from.value == v2.value && e.to.value == v1.value)) {
                return true;
            }
        }
        return false;
    }

    public int outDegree(Vertex v) {
        int degree = 0;
        for (Edge e : edges)
            if (e.from.value == v.value)
                degree++;
        return degree;
    }

    public int intDegree(Vertex v) {
        int degree = 0;
        for (Edge e : edges)
            if (e.to.value == v.value)
                degree++;
        return degree;
    }

    public void removeVertex(Vertex vertex) {
        for (Vertex v : vertices) {
            if (v.value == vertex.value) {
                vertices.remove(vertex.ID);
                return;
            }
        }

        System.out.println("There is no " + vertex.value + " vertex!!!");
    }

    public void removeEdge(Edge edge) {
        for (Edge e : edges) {
            if (e.from.value == edge.from.value && e.to.value == edge.to.value && edge.weight == e.weight) {
                edges.remove(edge);
            }
        }
        System.out.println("There is no such edge!!!");
    }

    public boolean findEdge(Edge edge) {
        for (Edge e : edges)
            if (edge.weight == e.weight && edge.from.value == e.from.value && edge.to.value == e.to.value)
                return true;
        return false;
    }

    public List<Edge> outgoingEdges(Vertex vertex) {
        List<Edge> outgoing = new ArrayList<>();
        for (Edge e : edges)
            if (e.from.value == vertex.value)
                outgoing.add(e);
        return outgoing;
    }

    public List<Edge> incomingEdges(Vertex vertex) {
        List<Edge> incoming = new ArrayList<>();
        for (Edge e : edges)
            if (e.to.value == vertex.value)
                incoming.add(e);
        return incoming;
    }

    public Edge getEdge(Vertex u, Vertex v) {
        for (Edge e : edges) {
            if (u.value == e.from.value && v.value == e.to.value)
                return e;
        }
        System.out.println("There is no such edge");
        return null;
    }

    public boolean empty() {
        return vertices.isEmpty();
    }

    public int numberOfVertices() {
        return vertices.size();
    }

    public int numberOfEdges() {
        return edges.size();
    }

    List<Vertex> allVertices() {
        return vertices;
    }

    List<Edge> allEdges() {
        return edges;
    }
}

